controllers/reconcilers/logging_installation/logging_installation_reconciler.go

```diff
@@ -0,0 +1,353 @@
+package logging_installation
+
+import (
+	"context"
+	"strings"
+
+	"github.com/go-logr/logr"
+	v14 "github.com/openshift/cluster-logging-operator/apis/logging/v1"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	v1 "github.com/redhat-developer/observability-operator/v3/api/v1"
+	"github.com/redhat-developer/observability-operator/v3/controllers/model"
+	"github.com/redhat-developer/observability-operator/v3/controllers/reconcilers"
+	v12 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type Reconciler struct {
+	client client.Client
+	logger logr.Logger
+}
+
+func NewReconciler(client client.Client, logger logr.Logger) reconcilers.ObservabilityReconciler {
+	return &Reconciler{
+		client: client,
+		logger: logger,
+	}
+}
+func (r *Reconciler) Cleanup(ctx context.Context, cr *v1.Observability) (v1.ObservabilityStageStatus, error) {
+	if cr.DescopedModeEnabled() {
+		return v1.ResultSuccess, nil
+	}
+
+	managed, err := r.checkLoggingSubscriptionForLabel(ctx)
+	if err != nil {
+		return v1.ResultFailed, err
+	}
+
+	if managed {
+		subscription := model.GetLoggingSubscription(cr)
+		err = r.client.Delete(ctx, subscription)
+		if err != nil && !errors.IsNotFound(err) {
+			return v1.ResultFailed, err
+		}
+
+		installedCsv, err := r.getInstalledLoggingCSVName(ctx)
+		if installedCsv == "" || err != nil {
+			return v1.ResultFailed, err
+		}
+
+		// Delete csv to uninstall
+		csv := &v1alpha1.ClusterServiceVersion{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      installedCsv,
+				Namespace: "openshift-logging",
+			},
+		}
+		err = r.client.Delete(ctx, csv)
+		if err != nil && !errors.IsNotFound(err) {
+			return v1.ResultFailed, err
+		}
```

Comment:
Think we should delete CSV before deleting subscription. getInstalledLoggingCSVName relies on gathering a subscriptionList so we need the subscription intact for this

Reply:
Hmm, I think you'll end up with a broken Operator if you delete the CSV at that point, but I could be wrong.

Reply:
Yes, good catch, I saw that in testing and changed the order but must have lost it somewhere along the way. I'll fix that. 

I think it makes sense to get the CSV name then delete the sub and then delete the CSV. I'm not sure if we delete the CSV while the sub still exists if there's a possibility the sub could recreate the CSV if there is some delay.

Reply:
Yeah good point about timing. Thanks Steven